src/ui/Sessions/backupservice.cpp

```diff
@@ -0,0 +1,149 @@
+#include <set>
+
+#include "include/Sessions/backupservice.h"
+#include "include/Sessions/persistentcache.h"
+#include "include/Sessions/sessions.h"
+
+#include "include/mainwindow.h"
+
+QTimer BackupService::s_autosaveTimer;
+std::vector<BackupService::WindowData> BackupService::s_backupWindowData;
+
+void BackupService::executeBackup() {
+    const auto& backupPath = PersistentCache::backupDirPath();
+
+    std::vector<WindowData> newData, unionOfData, savedData;
+
+    // Fill newData with up-to-date window data
+    for (const auto& wnd : MainWindow::instances()) {
+        WindowData wd;
+        wd.ptr = wnd;
+        wnd->topEditorContainer()->forEachEditor([&wd](int,int,EditorTabWidget*,Editor* ed) {
+            wd.editors.push_back( std::make_pair(ed, ed->getHistoryGeneration()) );
+            return true;
+        });
+        newData.push_back(std::move(wd));
+    }
+
+    // Set unionOfData to set union of old and new window data
+    std::set_union(s_backupWindowData.begin(), s_backupWindowData.end(),
+                   newData.begin(), newData.end(),
+                   std::back_inserter(unionOfData));
+
+    for (const auto& item : unionOfData) {
+        const auto oldIter = std::find(s_backupWindowData.begin(), s_backupWindowData.end(), item);
+        const auto newIter = std::find(newData.begin(), newData.end(), item);
+
+        const bool isInOld = oldIter != s_backupWindowData.end();
+        const bool isInNew = newIter != newData.end();
+
+        if (!isInNew) {
+            // These windows have been closed by the user. Remove their session caches.
+            const auto ptrToInt = reinterpret_cast<uintptr_t>(item.ptr);
+            const QString cachePath = backupPath + QString("/window_%1").arg(ptrToInt);
+            QDir(cachePath).removeRecursively();
+            continue;
+        }
+
+        if (isInOld && oldIter->isFullyEqual(*newIter)) {
+            // These windows are unchanged. Don't save them
+            savedData.push_back(*newIter);
+            continue;
+        }
+
+        // At this point the window is either new (!isInOld && isInNew) or changed (isInOld && !isFullyEqual(new,old)).
```

Comment:
Could you put a Q_ASSERT here?

Reply:
You mean for the conditions inside the comments? They're only meant for people to follow the code logic. The code itself doesn't care which one is true.

The key point is that `isInNew==true` at this point, meaning the current item can be saved. This is always the case since the check in line 40 skips over items where `isInNew==false`.